File: fortran/bench/mandel.f
C     Scalar-double-heavy kernel: a Mandelbrot-style iteration where
C     every hot value is a DOUBLE PRECISION scalar local, which is the
C     shape array-based LINPACK does not exercise.
      PROGRAM DSCAL2
      DOUBLE PRECISION X, Y, ZR, ZI, T, S
      INTEGER I, J, K, N
      N = 0
      DO 30 J = 1, 200
         Y = DBLE(J) / 100.0D0 - 1.0D0
         DO 20 I = 1, 200
            X = DBLE(I) / 100.0D0 - 1.5D0
            ZR = 0.0D0
            ZI = 0.0D0
            DO 10 K = 1, 60
               T = ZR*ZR - ZI*ZI + X
               ZI = 2.0D0*ZR*ZI + Y
               ZR = T
               S = ZR*ZR + ZI*ZI
               IF (S .GT. 4.0D0) GO TO 20
   10       CONTINUE
            N = N + 1
   20    CONTINUE
   30 CONTINUE
      IF (N .NE. 15756) STOP 1
      STOP 0
      END

File: fortran/bench/mandel.c
/* Mirror of mandel.f: every hot value is a double scalar local. */
int main(void) {
    double x, y, zr, zi, t, s;
    int i, j, k, n = 0;
    for (j = 1; j <= 200; j++) {
        y = (double)j / 100.0 - 1.0;
        for (i = 1; i <= 200; i++) {
            x = (double)i / 100.0 - 1.5;
            zr = 0.0; zi = 0.0;
            for (k = 1; k <= 60; k++) {
                t = zr*zr - zi*zi + x;
                zi = 2.0*zr*zi + y;
                zr = t;
                s = zr*zr + zi*zi;
                if (s > 4.0) goto next;
            }
            n++;
        next: ;
        }
    }
    return (n != 15756) ? 1 : 0;
}
